integrations/pjrt/src/iree_pjrt/common/api_impl.cc

```diff
@@ -852,6 +914,127 @@ iree_status_t DeviceInstance::HostBufferToDeviceZeroDim(
+    return ss.str();
+  };
+
+  auto arrayBuilder = [](const int64_t* vals, int64_t sz) {
+    std::stringstream ss;
+    ss << " {permutation = dense<[" << vals[0];
+    for (int i = 1; i < sz; ++i) ss << ", " << vals[i];
+    ss << "]> : tensor<" << sz << "xi64>}";
+    return ss.str();
+  };
+
+  auto broadcastBuilder = [](int64_t sz) {
+    std::stringstream ss;
+    ss << "{broadcast_dimensions = dense<[0";
+    for (int i = 1; i < sz; ++i) ss << ", " << i;
+    ss << "]> : tensor<" << sz << "xi64>}";
+    return ss.str();
+  };
+
+  const char* mlir_ty;
+  IREE_RETURN_IF_ERROR(
+      PJRTApiConverter::MapElementTypeToMlirType(element_type, &mlir_ty));
+
+  auto input_ty = typeBuilder(input_dims, num_dims, mlir_ty);
+  auto transpose_ty = typeBuilder(transpose_dims.data(), num_dims, mlir_ty);
+  auto output_ty = typeBuilder(output_dims, num_dims, mlir_ty);
+  auto perms_str = arrayBuilder(perms, num_dims);
+  auto broadcast_str = broadcastBuilder(num_dims);
+
+  std::stringstream program_builder;
+  program_builder << "func.func @main(\%arg0 : " << input_ty << ") -> ("
+                  << output_ty << ") {\n";
+  program_builder << "  %0 = \"stablehlo.transpose\"(\%arg0) " << perms_str
+                  << " : (" << input_ty << ") -> " << transpose_ty << "\n";
+  program_builder << "  %1 = \"stablehlo.broadcast_in_dim\"(%0) "
+                  << broadcast_str << " : (" << transpose_ty << ") -> "
+                  << output_ty << "\n";
+  program_builder << "  return %1 : " << output_ty << "\n";
+  program_builder << "}\n";
+
+  std::string transpose_program = program_builder.str();
+
+  // Create an on stack program:
+  PJRT_Program program;
+  program.code = transpose_program.data();
+  program.code_size = transpose_program.size();
+  program.format = kMlirFormat.data();
+  program.format_size = kMlirFormat.size();
+
+  // Compile program and check for errors:
+  LoadedExecutableInstance* executable;
+  auto* error = this->client().Compile(&program, /**options,*/ &executable);
+  if (error) {
+    auto errinst = ErrorInstance::FromError(error);
+    auto ret = iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                                "transposition program failed to build");
+    delete errinst;
+    return ret;
+  }
+
+  PJRT_Buffer* input = *buffer;
+  PJRT_Buffer** input_list = &input;
+
+  PJRT_Buffer* output;
+  PJRT_Buffer** output_list = &output;
+  PJRT_Event* event;
+
+  // Build the execution arguments for transposing the loaded memory:
+  PJRT_LoadedExecutable_Execute_Args execute_args;
+  memset(&execute_args, 0, sizeof(execute_args));
+
+  PJRT_ExecuteOptions execute_options;
+  memset(&execute_options, 0, sizeof(execute_options));
+  execute_args.executable = *executable;
+  execute_args.options = &execute_options;
+  execute_args.argument_lists = &input_list;
+  execute_args.output_lists = &output_list;
+  execute_args.num_devices = 1;
+  execute_args.num_args = 1;
+  execute_args.execute_device = nullptr;
```

Comment:
Don't you have to set the same device as the computation would run in here?

Reply:
Currently we can't specify the compute device. Not really sure what work needs to be done there.